qualtran/bloqs/qrom.py

```diff
@@ -0,0 +1,71 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import Dict, Tuple, Union
+
+import cirq
+import numpy as np
+from attrs import frozen
+from cirq_ft import QROM as CirqQROM
+
+from qualtran import Bloq, CompositeBloq, Register, Signature
+from qualtran.cirq_interop import CirqQuregT, decompose_from_cirq_op
+
+
+@frozen
+class QROM(Bloq):
+    """Gate to load data[l] in the target register when the selection stores an index l.
+
+    In the case of multi-dimensional data[p,q,r,...] we use multiple named
+    selection registers [p, q, r, ...] to index and load the data.
+
+    Args:
+        selection_bitsizes: The number of bits used to represent each selection register
+            corresponding to the size of each dimension of the array. Should be
+            the same length as the shape of each of the datasets.
+        data_bitsizes: The number of bits used to represent the data
+            registers. This can be deduced from the maximum element of each of the
+            datasets. Should be of length len(data), i.e. the number of datasets.
+        data_ranges: Tuple specifying the ranges of each axis of the data to be loaded.
+        cvs: The control values for the gate. Defaults to no controls.
+    """
+
+    selection_bitsizes: Tuple[int, ...]
+    data_bitsizes: Tuple[int, ...]
+    data_ranges: Tuple[int, ...]
+    cvs: Tuple[int, ...] = ()
```

Comment:
QROM should take the data to be loaded as input? What's the logic for not specifying the data? You want to create a placeholder bloq that has the right signature but don't care too much about the actual underlying data?

Reply:
Right yeah, maybe I'm working off old information, but do we expect bloqs to be hashable?

Reply:
Ideally this wouldn't be the case and we could test the correctness on the qualtran side too. Given the importance of qrom for other subroutines it's better to be able to check for correctness, so I'll make it mirror cirq_ft (in that it expects the actual data).

Reply:
Yes, Bloqs should be hashable.

Reply:
numpy arrays not being hashable lead us to use cirq.value_equality in cirq_ft, I wasn't sure what the appropriate thing to do in qualtran.

Reply:
Ok, after talking to Matt I think it's best to go the cirq.value_equality route (implement the necessary __hash__ etc routines and don't compare on the data for the moment), just to make it a bit more transparent. I will open an issue to track this. The alternative would be to just have a bloq to count resources, but I think it would be nice to be able to rely on qrom for some testing.